[DTensor] Support user-supplied Generator for random ops

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159933

If the user provides a generator kwarg to a random op (e.g.
nn.init.uniform_(..., generator=my_generator)), we can still advance
that generator's state in a SPMD-global way so that each local-tensor
gets appropriate values and the generator advances to the same state as
if it had operated on the full tensor.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

wconstab committed:
[DTensor] Support user-supplied Generator for random ops

If the user provides a generator kwarg to a random op (e.g.
nn.init.uniform_(..., generator=my_generator)), we can still advance
that generator's state in a SPMD-global way so that each local-tensor
gets appropriate values and the generator advances to the same state as
if it had operated on the full tensor.

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -197,8 +197,19 @@ def dispatch(
+                assert maybe_user_generator is None or isinstance(
+                    maybe_user_generator, torch.Generator
+                )
+                # maybe_user_generator = None
```
remove comment

XilunWu reviewed (commented):
Overall LGTM. Let me know if you need quick unblock.

We need to explicitly tell users to initialize the generator passed in with the same seed,  or we add this functionality to our `manual_seed` API but users still need to be aware of this.

wconstab commented on torch/distributed/tensor/_random.py:
```diff
@@ -191,7 +193,17 @@ def _manual_seed(self, parallel_seed: int) -> None:
+            # not because we need to keep a copy of it but because its the easiest way to make it work with the
+            # existing set/get APIs
+            g_name = str(id(generator))
+            if g_name not in self.rng_states:
+                self.rng_states[g_name] = generator.get_state()
```
I think I like the behavior this way- I don't want to introduce a collective on every op where a user supplies an RNG. 

I will add documentation to dtensor docs stating that it is the user's responsibility to ensure the passed generator has the same state on every spmd rank.

lucasb-eyer replied:
the original feature request comes from @akolesnikoff and myself. We noticed the issue specifically because we were wondering about and comparing behaviour of passing in same vs different seeded RNGs across ranks. So at least from our perspective, yes, we are intentional.

wconstab committed:
Update on "[DTensor] Support user-supplied Generator for random ops"


If the user provides a generator kwarg to a random op (e.g.
nn.init.uniform_(..., generator=my_generator)), we can still advance
that generator's state in a SPMD-global way so that each local-tensor
gets appropriate values and the generator advances to the same state as
if it had operated on the full tensor.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Support user-supplied Generator for random ops"


If the user provides a generator kwarg to a random op (e.g.
nn.init.uniform_(..., generator=my_generator)), we can still advance
that generator's state in a SPMD-global way so that each local-tensor
gets appropriate values and the generator advances to the same state as
if it had operated on the full tensor.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta

[ghstack-poisoned]

XilunWu commented on test/distributed/tensor/test_random_ops.py:
```diff
@@ -87,6 +87,39 @@ def test_init_ops(self):
+            (2, 3), device_mesh=device_mesh, placements=[Shard(0)]
+        )
+        for i in range(2):
+            print(f"{i=}")
```
this can be removed since we have 'i' in assert error

wconstab committed:
Update on "[DTensor] Support user-supplied Generator for random ops"


If the user provides a generator kwarg to a random op (e.g.
nn.init.uniform_(..., generator=my_generator)), we can still advance
that generator's state in a SPMD-global way so that each local-tensor
gets appropriate values and the generator advances to the same state as
if it had operated on the full tensor.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta

[ghstack-poisoned]

XilunWu reviewed (approved):
this API change lgtm

wanchaol commented on torch/distributed/tensor/_random.py:
```diff
@@ -191,7 +193,17 @@ def _manual_seed(self, parallel_seed: int) -> None:
+            # This is a little hacky, but for any user-passed generator, we store its state under a unique key,
+            # not because we need to keep a copy of it but because its the easiest way to make it work with the
+            # existing set/get APIs. We also ensure we remove it from rng_states after each _distribute_region.
+            g_name = "user-passed-generator"
```
I wonder if it could just be str(generator) as the key?

wanchaol reviewed (approved):
lgtm
